### PR TITLE
Get annotation instructions for a given id

### PR DESCRIPTION
**Pull request by Qining**

### CODE REVIEW

File: source/opt/def_use_manager.cpp [hunk 1]
```diff
@@ -85,6 +87,24 @@ UseList* DefUseManager::GetUses(uint32_t id) {
   return &iter->second;
 }
 
+const UseList* DefUseManager::GetUses(uint32_t id) const {
```
Comment (antiagainst) [inline source/opt/def_use_manager.cpp]: `const &`

File: source/opt/def_use_manager.h [hunk 2]
```diff
@@ -72,6 +73,9 @@ class DefUseManager {
   // Returns the use instructions for the given |id|. If there is no uses of
   // |id|, returns nullptr.
   UseList* GetUses(uint32_t id);
+  const UseList* GetUses(uint32_t id) const;
+  // Returns a set of annotation instrunctions whose target is the given |id|.
+  std::unordered_set<ir::Instruction*> GetAnnotations(uint32_t id) const;
```
Comment (antiagainst) [inline source/opt/def_use_manager.h]: Why `unordered_set` instead of a plain `vector`?

File: source/opt/def_use_manager.cpp (hunk 1 above)
Comment (Qining) [inline source/opt/def_use_manager.cpp]: that way what should be return if the id is not found?

Comment (antiagainst) [inline source/opt/def_use_manager.cpp]: Never mind.

File: test/opt/test_def_use.cpp [hunk 3]
```diff
@@ -1414,4 +1414,41 @@ INSTANTIATE_TEST_CASE_P(
       }));
 // clang-format on
 
+TEST(DefUseManagerSingleTest, GetAnnotations) {
+  // Code
+  const char* code =
+      "OpDecorate %1 Block "
+      "OpDecorate %1 RelaxedPrecision "
```
Comment (antiagainst) [inline test/opt/test_def_use.cpp]: We should also add some debug instructions here.

Comment (Qining) [inline test/opt/test_def_use.cpp]: One debug instruction is added in 92d6fd7. The debug instruction won't be counted as an annotation instruction.

File: test/opt/test_def_use.cpp [hunk 4]
```diff
@@ -1414,4 +1414,51 @@ INSTANTIATE_TEST_CASE_P(
...
+  EXPECT_EQ(2u, annotations_of_1.size());
+  EXPECT_EQ(0u, annotations_of_2.size());
+  EXPECT_EQ(0u, annotations_of_3.size());
+  EXPECT_EQ(0u, annotations_of_4.size());
+
+  using DecorationSet = std::unordered_set<SpvDecoration, std::hash<uint32_t>>;
+  DecorationSet actual_decorations;
+  DecorationSet expected_decorations;
+
+  for (auto* anno_inst : annotations_of_1) {
```
Comment (antiagainst) [inline test/opt/test_def_use.cpp]: I think you can directly check the two elements in the vector...

File: test/opt/test_def_use.cpp [hunk 5]
```diff
@@ -1414,4 +1414,51 @@ INSTANTIATE_TEST_CASE_P(
       }));
 // clang-format on
 
+TEST(DefUseManagerSingleTest, GetAnnotations) {
+  // Code
+  const char* code =
+      "OpName %1 \"struct_type\" "
+      "OpName %3 \"int_type\" "
+      "OpDecorate %1 Block "
+      "OpDecorate %1 RelaxedPrecision "
```
Comment (antiagainst) [inline test/opt/test_def_use.cpp]: DecorationGroup

Comment (Qining) [inline test/opt/test_def_use.cpp]: Just realized we have DisassembleInst() function available. Changed back to use parameterized tests in 5f94bfb

File: source/opt/def_use_manager.h [hunk 6]
```diff
@@ -72,6 +72,14 @@ class DefUseManager {
   // Returns the use instructions for the given |id|. If there is no uses of
   // |id|, returns nullptr.
   UseList* GetUses(uint32_t id);
+  const UseList* GetUses(uint32_t id) const;
+  // Returns the annotation instrunctions which are a directly use of the given
```
Comment (antiagainst) [inline source/opt/def_use_manager.h]: direct

File: source/opt/def_use_manager.h [hunk 7]
```diff
@@ -72,6 +72,14 @@ class DefUseManager {
...
+  // |id|. This means when the decorations are applied through decoration
+  // group(s), this function will just return the OpGroupDecorate
+  // instrcution(s) which refer to the given id as an operand. The OpDecorate
+  // instructions which decorates the decoration group will not be in the
```
Comment (antiagainst) [inline source/opt/def_use_manager.h]: which decorate

File: source/opt/def_use_manager.h [hunk 8]
```diff
@@ -72,6 +72,14 @@ class DefUseManager {
...
+  // return vector.
```
Comment (antiagainst) [inline source/opt/def_use_manager.h]: will not be returned.

**Comment by antiagainst** [comment]
+2. Just two grammar issues.

File: source/opt/def_use_manager.h (hunk 8 above)
Comment (Qining) [inline source/opt/def_use_manager.h]: Fixed in e95acf1

File: source/opt/def_use_manager.h (hunk 7 above)
Comment (Qining) [inline source/opt/def_use_manager.h]: fixed in e95acf1

File: source/opt/def_use_manager.h (hunk 6 above)
Comment (Qining) [inline source/opt/def_use_manager.h]: fixed in e95acf1

**Comment by Qining** [comment]
Both CI passed, squashed.

**Pull request merged**